File: packages/core/src/platform/services/agentService.ts
import type {
  AgentDefinition,
  Page,
  PageQuery,
  CreateAgentInput,
} from "@agently/contracts";
import { ids } from "@agently/contracts";
import type { PlatformDeps } from "./context.js";
import { paginate } from "../util.js";

export interface AgentService {
  list(query: PageQuery): Page<AgentDefinition>;
  create(input: CreateAgentInput): AgentDefinition;
}

export function createAgentService(deps: PlatformDeps): AgentService {
  const { repos, clock } = deps;
  return {
    list(query) {
      return paginate(repos.agents.all(), query);
    },
    create(input) {
      const now = clock.iso();
      const agent: AgentDefinition = {
        id: ids.agentDefinition(),
        workspaceId: repos.workspace.get().id,
        name: input.name,
        role: input.role,
        model: input.model,
        description: input.description,
        tools: input.tools,
        config: input.config,
        createdAt: now,
        updatedAt: now,
      };
      return repos.agents.insert(agent);
    },
  };
}

File: apps/api/internal/services/agent_service.go
package services

import (
	"github.com/agently/api/internal/domain"
	"github.com/agently/api/internal/domain/validate"
	"github.com/agently/api/internal/platform"
)

// AgentService lists and creates reusable agent definitions.
type AgentService struct {
	deps Deps
}

func NewAgentService(deps Deps) *AgentService {
	return &AgentService{deps: deps}
}

func (s *AgentService) List(query domain.PageQuery) domain.Page[domain.AgentDefinition] {
	return platform.Paginate(s.deps.Repos.Agents.All(), query)
}

func (s *AgentService) Create(input validate.CreateAgentInput) (domain.AgentDefinition, error) {
	now := domain.Timestamp(s.deps.Clock.ISO())
	agent := domain.AgentDefinition{
		ID: domain.NewAgentDefinitionId(), WorkspaceID: s.deps.Repos.Workspace.Get().ID,
		Name: input.Name, Role: input.Role, Model: input.Model, Description: input.Description,
		Tools: input.Tools, Config: input.Config, CreatedAt: now, UpdatedAt: now,
	}
	return s.deps.Repos.Agents.Insert(agent), nil
}
